import os

import pandas as pd
from snakemake.utils import min_version

min_version("6.0")

configfile: "config/config.yaml"

os.makedirs("snakemake_tmp", exist_ok=True)

#### Parse sample table ########################################################
sampletsv = pd.read_csv(config['sampletsv'], sep="\t", index_col=[0])
sampletsv.index = sampletsv.index.astype(str)
SAMPLES = sampletsv.index.tolist()
################################################################################

#### Include sub-workflows #####################################################

include: "rules/assembly.smk"
include: "rules/alignment.smk"
include: "rules/consensus_correction.smk"
include: "rules/rename_sort_contigs.smk"
include: "rules/finalise_alignments.smk"
#include: "rules/quality_summary.smk"
#include: "rules/prokka.smk"
#include: "rules/pydamage.smk"

################################################################################

rule all:
    input:
        expand("{resultdir}/consensus_correction/{assembler}/{sample}_contigs.fasta.gz", resultdir=[config['resultdir']], assembler=[config['assembler']], sample=SAMPLES),
        expand("{resultdir}/alignment/{assembler}/{sample}.sorted.dedup.bam.bai", resultdir=[config['resultdir']], assembler=[config['assembler']], sample=SAMPLES)

rule validate_sampletsv:
    output:
        touch("{tmpdir}/sampletsv.validated")
    message: "Validate the correctness of the input table"
    params:
        sampletsv = lambda wildcards: config['sampletsv'],
        assembler = lambda wildcards: config['assembler'],
        readcorrection = lambda wildcards: config['readcorrection']
    script:
        "scripts/validate_sampletsv.py"
